modules/renter/contractor/contracts.go

```diff
@@ -0,0 +1,445 @@
+	// If the host is not in the hostdb, then the host is missing.
+	host, exists := c.hdb.Host(contract.HostPublicKey)
+	if !exists {
+		return true
+	}
+
+	// Sanity check - ScanHistory should always be ordered from oldest to
+	// newest.
+	if build.DEBUG && !sort.IsSorted(host.ScanHistory) {
+		sort.Sort(host.ScanHistory)
+		build.Critical("host's scan history was not sorted")
+	}
+
+	// Consider a host offline if:
+	// 1) The host has been scanned enough times.
+	// 2) The most recent scans have all failed.
+	// 3) The time between the most recent scan and the last successful scan
+	//    (or first scan) has exceeded the acceptable amount.
+	numScans := len(host.ScanHistory)
+	if numScans < missingMinScans {
+		// Not enough data to make a fair judgment.
+		return false
+	}
+	recent := host.ScanHistory[numScans-missingMinScans:]
+	for _, scan := range recent {
+		if scan.Success {
+			// One of the recent scans succeeded.
+			return false
+		}
+	}
+
+	// Initialize window bounds.
+	lastScan := host.ScanHistory[0].Timestamp
+	lastSuccess := host.ScanHistory[numScans-1].Timestamp
+	// Iterate from newest-oldest, seeking to last successful scan.
+	for i := numScans - 1; i >= 0; i-- {
+		if host.ScanHistory[i].Success {
+			lastSuccess = host.ScanHistory[i].Timestamp
+			break
+		}
+	}
+	return lastScan.Sub(lastSuccess) >= missingWindow
+}
+
+// managedMarkBadContracts will go through the contractors set of contracts and mark
+// any of the contracts which are no longer performing well.
+func (c *Contractor) managedMarkBadContracts() {
+	c.mu.Lock()
+	desiredHosts := c.allowance.Hosts
+	c.mu.Unlock()
+	// The hosts will be compared against the hosts in the hostdb to determine
+	// whether they have an acceptable score or if they should be replaced.
+	// Determine what counts as an acceptable score.
+	hosts := c.hdb.RandomHosts(int(desiredHosts), nil)
+	// Find the host of the bunch with the lowest score.
+	var lowestScore types.Currency
+	if len(hosts) > 0 {
+		// Get the score of the first host, as a baseline.
+		lowestScore = c.hdb.ScoreBreakdown(hosts[0]).Score
+		// Find the lowest score overall.
+		for _, host := range hosts {
+			if c.hdb.ScoreBreakdown(host).Score.Cmp(lowestScore) < 0 {
+				lowestScore = c.hdb.ScoreBreakdown(host).Score
+			}
+		}
+	}
+	// Adjust the lowest score further down. This gives the hosts wiggle room to
+	// not need to be exactly the best performers in order to be acceptable.
+	lowestScore = lowestScore.Div64(badScoreForgiveness)
+
+	// Now that we've finished interacting with external packages, lock the
+	// contractor and begin marking bad contracts.
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// Check each host for various conditions that would cause them to be
+	contracts := make([]modules.RenterContract, 0, len(c.contracts))
+	for _, contract := range c.contracts {
+		contracts = append(contracts, contract)
+	}
+	for _, contract := range contracts {
```

Comment:
Right, Go won't let you do `m["foo"].bar = 2` because `"foo"` may not be in the map. Interestingly, it is legal to do `m["baz"] += 2` because that's short for `m["baz"] = m["baz"] + 2`, and `m["baz"]` will return `0` if the key doesn't exist, so the expression is still well-defined.

Anyway, couldn't you still do:
```go
for _, contract := range contracts {
    contract.Field++
    contracts[contract.ID] = contract
}
```